test/unit/OrderedMsgChain.test.js

```diff
@@ -0,0 +1,157 @@
+        util.add(msg1)
+        util.add(msg2)
+        util.add(msg5)
+    })
+    it('handles unordered messages in order', () => {
+        const received = []
+        util = new OrderedMsgChain('publisherId', 'msgChainId', (msg) => {
+            received.push(msg)
+        }, () => {})
+        util.add(msg1)
+        util.add(msg2)
+        util.add(msg5)
+        util.add(msg3)
+        util.add(msg4)
+        assert.deepStrictEqual(received, [msg1, msg2, msg3, msg4, msg5])
+    })
+    it('handles unchained messages in the order in which they arrive if they are newer', () => {
+        const m2 = createMsg(4, 0)
+        const m3 = createMsg(17, 0)
+        const m4 = createMsg(7, 0)
+        const received = []
+        util = new OrderedMsgChain('publisherId', 'msgChainId', (msg) => {
+            received.push(msg)
+        }, () => {})
+        util.add(msg1)
+        util.add(m2)
+        util.add(m3)
+        util.add(m4)
+        assert.deepStrictEqual(received, [msg1, m2, m3])
+    })
+    it('does not call the gap handler (scheduled but resolved before timeout)', () => {
+        const received = []
+        util = new OrderedMsgChain('publisherId', 'msgChainId', (msg) => {
+            received.push(msg)
+        }, () => {
+            throw new Error('Unexpected gap')
+        }, 10000)
+        util.add(msg1)
+        util.add(msg5)
+        util.add(msg4)
+        util.add(msg3)
+        util.add(msg2)
+        assert.deepStrictEqual(received, [msg1, msg2, msg3, msg4, msg5])
+    })
+    it('call the gap handler MAX_GAP_REQUESTS times and then throws', (done) => {
+        const clock = sinon.useFakeTimers()
+        let counter = 0
+        try {
+            util = new OrderedMsgChain('publisherId', 'msgChainId', () => {}, (from, to, publisherId, msgChainId) => {
+                assert.strictEqual(from.timestamp, msg1.getMessageRef().timestamp)
+                assert.strictEqual(from.sequenceNumber, msg1.getMessageRef().sequenceNumber + 1)
+                assert.deepStrictEqual(to, msg3.prevMsgRef)
+                assert.strictEqual(publisherId, 'publisherId')
+                assert.strictEqual(msgChainId, 'msgChainId')
+                counter += 1
+                clock.tick(5000)
+            }, 5000)
+            util.add(msg1)
+            util.add(msg3)
+            clock.tick(5000)
+        } catch (e) {
+            assert.strictEqual(e.message, 'Failed to fill gap between [1,1] and [2,0] for publisherId-msgChainId'
+                + ` after ${OrderedMsgChain.MAX_GAP_REQUESTS} trials`)
+            assert.strictEqual(counter, OrderedMsgChain.MAX_GAP_REQUESTS)
+            clock.restore()
+            done()
+        }
+    })
+    it('handles unordered messages in order (large randomized test)', () => {
+        const expected = [msg1]
+        let i
+        for (i = 2; i <= 1000; i++) {
+            expected.push(createMsg(i, 0, i - 1, 0))
+        }
+        const shuffled = shuffle(expected)
```

Comment:
It would be a good idea to provide the `shuffle` function a fixed seed if possible so that the test it deterministic.

Reply:
Why does the test need to be deterministic?
I made it random on purpose: I thought that having a different input each time the test is run makes it way more robust than a deterministic test. No matter the order of the input messages, we always expect the same order of output messages

Reply:
Reproducibility of failing tests is important. Let's say I'm working on something completely different, run all tests, and then this test, on some specific outcome of shuffle, happens to fail. It won't really help me fix the issue because the next run of this test passes / behaves differently.

Pretty good discussion about this here: https://softwareengineering.stackexchange.com/questions/221632/testing-deterministic-or-non-deterministic

Basically what it boils down to is that tests needs to be reproducible. And that requires the inputs to the process to be known. Some ideas on how this can be achieved
1) use a fixed seed
2) try running multiple times with a set of random pre-selected seeds
3) console.log the order used and make it easy for the user to run the same scenario using that order